core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -930,6 +962,69 @@ public void onFailure(Exception shardFailedError) {
         }
     }
 
+    /** a wrapper class to encapsulate a request when being sent to a specific allocation id **/
+    final class RequestWithAllocationID<R extends TransportRequest> extends TransportRequest {
```

Comment:
to me `ConcreteShardRequest` is meaningless unless you know about this PR and its context, which all of us will soon forget. If you feel `ConcreteShardRequest` is a much better name, I will happily make you happy :)